Sources/System/Sockets/SocketDescriptor+Messages.swift

```diff
@@ -0,0 +1,333 @@
+    }
+  }
+
+  public var startIndex: Index { Index(_offset: 0) }
+  public var endIndex: Index { Index(_offset: _endOffset) }
+  public var isEmpty: Bool { _endOffset == 0 }
+
+  public var count: Int {
+    _withUnsafeBytes { buffer in
+      var count = 0
+      var p = buffer.baseAddress!
+      let end = buffer.baseAddress! + buffer.count
+      while p <= end - _headerSize {
+        count += 1
+        let header = p.assumingMemoryBound(to: CInterop.CMsgHdr.self)
+        let length = Int(header.pointee.cmsg_len)
+        p += length
+      }
+      return count
+    }
+  }
+
+  public func index(after i: Index) -> Index {
+    precondition(i._offset != _endOffset, "Can't advance past endIndex")
+    precondition(i._offset >= 0 && i._offset + _headerSize <= _endOffset,
+                 "Invalid index")
+    let delta: Int = _withUnsafeBytes { buffer in
+      let p = (buffer.baseAddress! + i._offset)
+        .assumingMemoryBound(to: CInterop.CMsgHdr.self)
+      return Int(p.pointee.cmsg_len)
+    }
+    if i._offset + delta + _headerSize > _endOffset {
+      return endIndex
+    }
+    return Index(_offset: i._offset + delta)
+  }
+
+  public subscript(position: Index) -> Message {
+    precondition(
+      position._offset >= 0 && position._offset + _headerSize <= _endOffset,
+      "Invalid index")
+    return Element(_base: self, offset: position._offset)
+  }
+}
+
+// @available(macOS 9999, iOS 9999, watchOS 9999, tvOS 9999, *)
+extension SocketDescriptor.ControlMessageBuffer.Message {
+  internal var _header: CInterop.CMsgHdr {
+    _base._withUnsafeBytes { buffer in
+      assert(_offset + _base._headerSize <= buffer.count)
+      let p = buffer.baseAddress! + _offset
+      let header = p.assumingMemoryBound(to: CInterop.CMsgHdr.self)
+      return header.pointee
+    }
+  }
+
+  public var level: CInt {
+    _header.cmsg_level
+  }
+
+  public var type: CInt {
+    _header.cmsg_type
+  }
+
+  public func withUnsafeBytes<R>(
+    _ body: (UnsafeRawBufferPointer) throws -> R
+  ) rethrows -> R {
+    try _base._withUnsafeBytes { buffer in
+      assert(_offset + _base._headerSize <= buffer.count)
+      let p = buffer.baseAddress! + _offset
+      let header = p.assumingMemoryBound(to: CInterop.CMsgHdr.self)
+      let length = Int(header.pointee.cmsg_len)
+      let data = p + _base._headerSize
+      let count = Swift.min(_offset + length, buffer.count)
+      return try body(UnsafeRawBufferPointer(start: data, count: count))
+    }
+  }
+}
+
+extension Optional where Wrapped == SocketDescriptor.ControlMessageBuffer {
+  internal func _withUnsafeBytes<R>(
+    _ body: (UnsafeRawBufferPointer) throws -> R
+  ) rethrows -> R {
+    guard let messages = self else {
+      return try body(UnsafeRawBufferPointer(start: nil, count: 0))
+    }
+    return try messages._withUnsafeBytes(body)
+  }
+}
+
+// @available(macOS 9999, iOS 9999, watchOS 9999, tvOS 9999, *)
+extension SocketDescriptor {
+  public func sendMessage(
```

Comment:
I would sink this logic into the `Result()` yielding one so that this is just a `Result->Error` wrapper that can be emitted into client. Then we can upgrade to typed throws if/when that ever happens. I think it's clearer too if it follows existing precedent.

Reply:
And putting this in `SocketOperations.swift` might be clearer too.